### PR TITLE
systemd-stage-1: Support for user shells

### PR DESCRIPTION
**Pull request by ElvishJerricco**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/config/users-groups.nix [hunk 1]
```diff
@@ -750,17 +758,20 @@ in {
...
           ${lib.concatStringsSep "\n" (lib.mapAttrsToList (n: { gid }: "${n}:x:${toString gid}:") config.boot.initrd.systemd.groups)}
         '';
+        "/etc/shells".text = lib.concatStringsSep "\n" (lib.unique (lib.mapAttrsToList (_: u: u.shell) config.boot.initrd.systemd.users)) + "\n";
       };
 
+      storePaths = [ "${pkgs.shadow}/bin/nologin" ];
+
       users = {
-        root = {};
+        root = { shell = lib.mkDefault "/bin/bash"; };
```
Comment (nikstur) [inline nixos/modules/config/users-groups.nix:774]: Isn't the "normal" or default case that you don't login into your initrd?  Wouldn't `nologin` be a  better default then?

Comment (ElvishJerricco) [reply nixos/modules/config/users-groups.nix:774]: Well, login is disabled via `/etc/shadow` by default. When `emergencyAccess` is set to something other than false, this needs to not be `nologin`

**Review by RaitoBezarius — APPROVED** [review]

**Review by nikstur — APPROVED** [review]

File: nixos/modules/system/boot/initrd-ssh.nix [hunk 2]
```diff
@@ -164,13 +164,12 @@ in
...
         '';
       }
-
-      {
-        assertion = config.boot.initrd.systemd.enable -> cfg.shell == null;
-        message = "systemd stage 1 does not support boot.initrd.network.ssh.shell";
-      }
     ];
 
+    warnings = lib.optional (config.boot.initrd.systemd.enable -> cfg.shell != null) ''
```
Comment (SuperSandro2000) [inline nixos/modules/system/boot/initrd-ssh.nix:169]: I don't think this is correct.

```
nix-repl> nixosConfigurations.hydrogen.config.boot.initrd.systemd.enable -> nixosConfigurations.hydrogen.config.boot.initrd.network.ssh.shell != null
true

nix-repl> nixosConfigurations.hydrogen.config.boot.initrd.network.ssh.shell
null

nix-repl> nixosConfigurations.hydrogen.config.boot.initrd.systemd.enable
false
```

Comment (ElvishJerricco) [reply nixos/modules/system/boot/initrd-ssh.nix:169]: I think you're right. When transitioning from assertion to warning, I switched the `==` with `!=`, which is good, but I didn't change `->` to `&&`.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/initrd-ssh.nix:169]: @SuperSandro2000 #266116

**Pull request merged**